### PR TITLE
Dynamodb case class with annotations (hashPK, rangePK) support

### PR DESCRIPTION
**Pull request by jwszolek**
PR adds annotations support to the case class definition entity. UT added. All tests are passing.

### CODE REVIEW

**Comment by jwszolek** [comment]
hey @seratch, could you please take a look? Not sure about your recommendation around reflection usage.

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 1]
```diff
@@ -64,40 +66,43 @@ case class Table(
   def put(hashPK: Any, attributes: SimplePk*)(implicit dynamoDB: DynamoDB): Unit = putItem(hashPK, attributes: _*)
   def put(hashPK: Any, rangePK: Any, attributes: SimplePk*)(implicit dynamoDB: DynamoDB): Unit = putItem(hashPK, rangePK, attributes: _*)
 
-  def putItem[E <: AnyRef](entity: E)(implicit dynamoDB: DynamoDB): Unit = {
-    val attrs = getAttrValuesToMap(entity, true)
+  def putItem[T: u.TypeTag](entity:T)(implicit dynamoDB: DynamoDB): Unit = {
+    val annotations = getterAnnotationsFromEntity(entity)
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:70]: Can you avoid using tuples for entries in the seq? Also, instead of accessing `._1` etc, using pattern matching should make your code more readable. 

For instance, the following code ...

```scala
    val faJoined = annotations.map(tup1 => fields.filter(tup2 => tup2._1 == tup1._1.toString)
      .map(tup2 => (tup1._1.toString, tup1._2.head.toString, {
        tup1._3.toString match {
          case "Int" => tup2._2.asInstanceOf[Int]
          case "String" => tup2._2.asInstanceOf[String]
        }
      }, tup1._3.toString)))
```

can be like this:

```scala
    val faJoined2: Seq[List[(String, String, Any, String)]] = annotations.map {
      case (name, annotations, sig) =>
        val found: Option[(String, AnyRef)] = fields.find { case (fieldName, _) => fieldName == name.toString }
        found match {
          case Some((_, getterValue)) =>
            List((
              name.toString,
              annotations.head.toString,
              {
                sig.toString match {
                  case "Int" => getterValue.asInstanceOf[Int]
                  case "String" => getterValue.asInstanceOf[String]
                }
              }, sig.toString))
          case None =>
            Nil
        }
      case _ => Nil
    }
```

The code is a bit longer but having explicit types and using pattern matching should make it easier-to-understand for everyone (at least for me). Does the `faJoined` really need to be nested seq here? Also, the code calls toString so many times. Probably, `getterAnnotationsFromEntity` can return more usable values.

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 2]
```diff
@@ -64,40 +66,43 @@ case class Table(
...
 
-    if (!keys.exists(k => k._1 == "hashPK") && keysRequired)
-      throw new Exception(s"Primary key is not defined for ${entity.getClass.getName}")
+  def putItem[E <: AnyRef](hashPK: Any, rangePK: Any, entity: E)(implicit dynamoDB: DynamoDB): Unit = {
+    val attrs = getterNamesFromEntity(entity)
+    dynamoDB.put(this, hashPK, rangePK, attrs: _*)
+  }
 
-    Map("keys" -> keys, "attributes" -> attrs)
+  private def getterAnnotationsFromEntity[T: u.TypeTag](entity: T) = {
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:100]: Let's be more conscious about the return type. Avoid handy tuples as much as possible for readability. Having a private class for the return type is also an option. Having a name for each field in a class is helpful for understanding the meaning of objects here.

**Review by seratch — CHANGES_REQUESTED** [review]
I'm not a big fan of annotations in Scala but it's okay to use reflection and annotations for this purpose. 

One thing I have to say is that the current implementation is not great enough. Can you make the code a bit easier-to-understand and optimal by following my suggestions in the comments?

File: README.md [hunk 3]
```diff
@@ -239,8 +239,14 @@ import awscala._, dynamodbv2._
 implicit val dynamoDB = DynamoDB.at(Region.Tokyo)
 
 case class Member(Name: String, Age: Int, Company: String)
-case class User(hashPK: Int, rangePK: String, Name: String, Age: Int, Company: String)
-
+case class TestMember(
+                 @hashPK
```
[the remark this answers is not in the document]
Comment (jwszolek) [reply README.md]: cool, will use it before pushing any changes into git

File: README.md [hunk 4]
```diff
@@ -239,8 +239,14 @@ import awscala._, dynamodbv2._
...
+                 id: Int,
+                 @rangePK
+                 Country: String,
```
[the remark this answers is not in the document]
Comment (jwszolek) [reply README.md]: done

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 1 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: thanks for the recommendations and advice! My code was updated. Let me know if any other changes are needed.

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 2 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: I have added `AnnotationMeta` case class. This seems to be more consistent now.

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 5]
```diff
@@ -64,40 +70,63 @@ case class Table(
...
-      dynamoDB.put(this, hashPK, attrs("attributes"): _*)
+  def putItem[T: u.TypeTag](entity: T)(implicit dynamoDB: DynamoDB): Unit = {
+    val annotations: List[AnnotationMeta] = getterAnnotationsFromEntity(entity)
+    val fields: List[(String, AnyRef)] = getterNamesFromEntity(entity)
+
+    val faJoined = annotations.map(a => {
+      val found: Option[(String, AnyRef)] = fields.find { case (fieldName, _) => fieldName == a.name }
+      found match {
+        case Some((_, getterValue)) =>
+          (
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:81]: can you avoid using a tuple here?

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 6]
```diff
@@ -64,40 +70,63 @@ case class Table(
...
+            a.name,
+            a.annotation, {
+              a.typeSignature match {
+                case "Int" => getterValue.asInstanceOf[Int]
+                case "String" => getterValue.asInstanceOf[String]
+              }
+            },
+            a.typeSignature)
+        case None => Nil
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:90]: If the element in this pattern is empty, you can use `None` instead of `Nil` here. And the `annotations.map` can be `annotations.flatMap`.

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 7]
```diff
@@ -64,40 +70,63 @@ case class Table(
...
-      val rangePK = attrs("keys").find(f => f._1 == "rangePK").get._2
-      dynamoDB.put(this, hashPK, rangePK, attrs("attributes"): _*)
-    } else if (attrs("keys").exists(x => x._1 == "hashPK")) {
-      val hashPK = attrs("keys").find(f => f._1 == "hashPK").get._2
-      dynamoDB.put(this, hashPK, attrs("attributes"): _*)
+  def putItem[T: u.TypeTag](entity: T)(implicit dynamoDB: DynamoDB): Unit = {
+    val annotations: List[AnnotationMeta] = getterAnnotationsFromEntity(entity)
+    val fields: List[(String, AnyRef)] = getterNamesFromEntity(entity)
+
+    val faJoined = annotations.map(a => {
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:77]: can you have an explicit type for `faJoined`? also, the name is not clear enough.

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 8]
```diff
@@ -64,40 +70,63 @@ case class Table(
...
+                case "Int" => getterValue.asInstanceOf[Int]
+                case "String" => getterValue.asInstanceOf[String]
+              }
+            },
+            a.typeSignature)
+        case None => Nil
+      }
+    })
+
+    val hashKey: Option[Any] = faJoined.asInstanceOf[List[(String, String, Any, String)]]
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:94]: please don't cast the type using `asInstanceOf`

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 9]
```diff
@@ -106,10 +135,14 @@ case class Table(
...
       .filterNot(m => Modifier.isStatic(m.getModifiers))
-      .filterNot(m => m.getParameterTypes.size > 0)
+      .filterNot(m => m.getParameterTypes.length > 0)
       .map(_.getName)
 
     methodNames.filter(m => fieldNames.contains(m))
+      .map(getterName => {
+        val value = obj.getClass.getDeclaredMethod(getterName).invoke(obj)
+        getterName -> value
+      }).toList
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:145]: Why did you add toList here? I don't see the necessity to use List over Seq here.

**Review by seratch — CHANGES_REQUESTED** [review]

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 9 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: fixed

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 8 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: done

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 7 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: changed to annotationFields

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 10]
```diff
@@ -64,40 +70,66 @@ case class Table(
...
-      val hashPK = attrs("keys").find(f => f._1 == "hashPK").get._2
-      dynamoDB.put(this, hashPK, attrs("attributes"): _*)
+  def putItem[T: u.TypeTag](entity: T)(implicit dynamoDB: DynamoDB): Unit = {
+    val annotations: List[AnnotationMeta] = getterAnnotationsFromEntity(entity)
+    val fields: Seq[(String, AnyRef)] = getterNamesFromEntity(entity)
+
+    val annotationFields:List[Option[AnnotationMeta]] = annotations.map(a => {
+      val found: Option[AnyRef] = fields
+        .find { case (fieldName, _) => fieldName == a.name }
+        .flatMap{ case(_, getterValue) => Some(getterValue) }
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:80]: it seems code formatter is not applied

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 11]
```diff
@@ -64,40 +70,66 @@ case class Table(
...
-      val rangePK = attrs("keys").find(f => f._1 == "rangePK").get._2
-      dynamoDB.put(this, hashPK, rangePK, attrs("attributes"): _*)
-    } else if (attrs("keys").exists(x => x._1 == "hashPK")) {
-      val hashPK = attrs("keys").find(f => f._1 == "hashPK").get._2
-      dynamoDB.put(this, hashPK, attrs("attributes"): _*)
+  def putItem[T: u.TypeTag](entity: T)(implicit dynamoDB: DynamoDB): Unit = {
+    val annotations: List[AnnotationMeta] = getterAnnotationsFromEntity(entity)
+    val fields: Seq[(String, AnyRef)] = getterNamesFromEntity(entity)
+
+    val annotationFields:List[Option[AnnotationMeta]] = annotations.map(a => {
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:77]: This can be `Seq[AnnotationMeta]` - you can do `flatMap` instead of `map`

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 12]
```diff
@@ -64,40 +70,66 @@ case class Table(
...
+      val found: Option[AnyRef] = fields
+        .find { case (fieldName, _) => fieldName == a.name }
+        .flatMap{ case(_, getterValue) => Some(getterValue) }
+
+      found match {
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:82]: if you want to do this, using `map` method rather than match expression can be simpler.

```
found.map { getterValue => /* the same code */ } // returns Option

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala [hunk 13]
```diff
@@ -64,40 +70,66 @@ case class Table(
...
+              case "Int" => getterValue.asInstanceOf[Int]
+              case "String" => getterValue.asInstanceOf[String]
+            }
+          }))
+        case None => None
+      }
+    })
+
+    val hashKey: Option[Any] = annotationFields
+      .find ( a => a.get.annotation.contains("hashPK") )
```
Comment (seratch) [inline dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala:98]: `a.get` shows the code smell

**Review by seratch — CHANGES_REQUESTED** [review]

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 12 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: hey @seratch  - thanks for the review. I guess map seems to work better in this case. Pushed into git. Please take a look

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 13 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: that's a general rule? avoid `get` where it is possible?

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 10 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: formatter applied

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 13 above)
Comment (seratch) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: You should never use the method in Scala. `Option.get` method can throw runtime exception `java.util.NoSuchElementException` if the value is `None`. Doing so completely eliminates the benefits of Option values.

**Review by seratch — APPROVED** [review]
OKay, thanks - I will modify some parts after merging this PR

File: dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala (hunk 13 above)
Comment (jwszolek) [reply dynamodb/src/main/scala/awscala/dynamodbv2/Table.scala]: gotcha, thanks!
